### ISSUE TITLE
Feature request: Detect double initialization of library

### ISSUE THREAD

**Issue by AndyBarron** [reduxjs/react-redux#1077]
**TL;DR: Emit warning if multiple copies of React Redux are imported**

The new React context API requires reference equality of `Context` instances. As a result, if two copies of `react-redux` are pulled in <sup>1</sup>, `ReactReduxContext.Consumer` will fail to locate a `store` instance and crash. The component tree ends up looking something like this:

```jsx
<ReactReduxContextA.Provider>
  <ReactReduxContextB.Consumer>
    {(value) => (
      <div>Will be null: { JSON.stringify(value) }</div>
    )}
  </ReactReduxContextB.Consumer>
</ReactReduxContextA.Provider>
```

This is 100% user error, but tracking it down can be time-consuming.<sup>2</sup> [Styled Components](https://github.com/styled-components/styled-components) helps detect this error with the following check<sup>3</sup>:

```js
/* Warning if there are several instances of styled-components */
if (
  process.env.NODE_ENV !== 'production' &&
  process.env.NODE_ENV !== 'test' &&
  typeof window !== 'undefined' &&
  typeof navigator !== 'undefined' &&
  typeof navigator.userAgent === 'string' &&
  navigator.userAgent.indexOf('Node.js') === -1 &&
  navigator.userAgent.indexOf('jsdom') === -1
) {
  window['__styled-components-init__'] = window['__styled-components-init__'] || 0;

  if (window['__styled-components-init__'] === 1) {
    // eslint-disable-next-line no-console
    console.warn(
      "It looks like there are several instances of 'styled-components' initialized in this application. " +
        'This may cause dynamic styles not rendering properly, errors happening during rehydration process ' +
        'and makes your application bigger without a good reason.\n\n' +
        'See https://s-c.sh/2BAXzed for more info.'
    );
  }

  window['__styled-components-init__'] += 1;
}
```

React Redux could more or less copy this approach to help out foolish people such as myself. :-)

<sup>1</sup> e.g. through a misconfigured Webpack bundle pulling in both CommonJS and ESM versions... Oops.
<sup>2</sup> Totally not speaking from personal experience.
<sup>3</sup> [Styled Components source code](https://github.com/styled-components/styled-components/blob/03e1fce9f71be592255cf377940f738f0297b6f5/src/base.js#L31-L54)

**Comment by markerikson** [comment]
Thanks, appreciate both the "error" report, and the example code!  We'll look into adding something like that.

**Comment by fauxsoup** [comment]
@AndyBarron - what was the error in your webpack config? I seem to be running into this issue with a minimal webpack configuration and am looking for a workaround that doesn't involve carrying my own context all over the place.

**Comment by SethDavenport** [comment]
This happens to me due to yarn link.

Our webpack config is a little special in the sense that:

* we have one top-level repo that contains all the webpack set up
* a bunch of different repos that different internal teams use to implement sections of our app. 

The store is configured by the top level repo, and the `<Provider>` block exists there; but stuff inside that block can be pulled in from a variety of the downstream repos.

Everything works fine in prod. But in devmode, teams 'yarn link' their repos into the toplevel one, and we end up with two copies of the lib.

In our case the fix was to add this to our webpack config:
```js
  resolve: {
    alias: {
       'react-redux': path.join(__dirname, 'node_modules', 'react-redux'),
     },
  },
```

However a warning message as described above would have helped me figure it out faster :D, so +1

**Comment by markerikson** [comment]
I don't think this is something we're going to ever get around to ourselves.  If someone would like to submit a PR to add it somehow we can take a look, though.
